### ISSUE TITLE
cartopy tiles in Orthographic projection miss some datas

### ISSUE THREAD

**Issue by ludwigVonKoopa** [SciTools/cartopy#1907]
### Description

When using tiles in Orthographic projection, i noticed missing data.

```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
from cartopy.io.img_tiles import Stamen

fig = plt.figure(figsize=(5,5), dpi=120)
ax = fig.add_subplot(1, 1, 1, projection=ccrs.Orthographic(-10, 60))
tiler = Stamen('terrain-background')

ax.add_image(tiler, 3)
```

![image](https://user-images.githubusercontent.com/49512274/136713451-6bed5746-0b10-46d5-866b-5ff216d9faeb.png)

I noticed white data at the top of the globe. It doesn't appear with precision < 3 

```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
from cartopy.io.img_tiles import Stamen

tiler = Stamen('terrain-background')
fig = plt.figure(figsize=(5,5), dpi=120)
for x in range(1,5):
    # print(x)
    ax = fig.add_subplot(2, 2, x, projection=ccrs.Orthographic(-10, 60))

    ax.add_image(tiler, x)
    ax.set_title(f"tile pre={x}")
```
![image](https://user-images.githubusercontent.com/49512274/136713554-ca0d4028-077c-4db9-b737-867a55b1fae7.png)

Is that something wanted ? 

Thanks

### Cartopy version

```python
import cartopy
cartopy.__version__
```
0.20.1

**Comment by greglucas** [comment]
It looks like the stamen-terrain doesn't go fully up to the pole. There is a band near the pole in PlateCarree as well.
```
ax = fig.add_subplot(1, 1, 1, projection=ccrs.PlateCarree())
ax.set_global()
```

I think this is unactionable on Cartopy's side.

**Comment by QuLogic** [comment]
I don't think this is referring to the pole, but the top of the circle; one tile that only shows a few pixels is not added.

**Comment by ludwigVonKoopa** [comment]
I checked this issue, i found the "problem".

When cartopy load the image, it load only the tiles who will be seen in the final images (i.e. skip tiles which are behind the earth).
This is the PlateCarree representation of the projected tiles loaded, it is more understandable : 

<details><summary>code to create figure</summary>

```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
from cartopy.io.img_tiles import Stamen
tiler = Stamen(style="terrain-background", cache="/tmp/tiles/")


fig = plt.figure(figsize=(10,8), dpi=120)

proj = ccrs.Orthographic(0, 45)

ax_pla = fig.add_subplot(2, 1, 1, projection=ccrs.PlateCarree(), frameon=True)
ax_ort = fig.add_axes([0., 0.32, 1, 0.3], projection=proj)

# this is what cartopy.mpl.geoaxes.GeoAxes.draw do for tiles computing
geom = ax_ort._get_extent_geom(tiler.crs)
img, extent, origin = tiler.image_for_domain(geom, 4)

for ax in [ax_pla, ax_ort]:
    ax.imshow(img, extent=extent, origin=origin, transform=tiler.crs)
    
ax_pla.set_title("tiles loaded for ccrs.Orthographic(0, 45) projection")
``` 
</details>

![image](https://user-images.githubusercontent.com/49512274/143691866-35dbb5a1-cf66-4590-84b5-92256cedd79d.png)

We see why there is no water sometimes on the border of the earth : tiles are not loaded.

To have a better representation, i plot the projection border on the plot : 

<details><summary>code to create figure</summary>

```python
import matplotlib.patches as mpatches
import shapely.geometry as sgeom

bounds_proj = sgeom.Polygon(proj.boundary)

for ax in [ax_pla, ax_ort]:
    c_bounds = ax.add_geometries([bounds_proj], crs=proj, color="green", alpha=0.3, zorder=40)
    
geo2 = mpatches.Patch(color='green', label="ortho boundaries")
ax_pla.legend(handles=[geo2], loc='lower right')
```
</details>

![image](https://user-images.githubusercontent.com/49512274/143691927-2483071b-55c4-41be-bbd3-1d1462959d8d.png)

We see white patch inside the green polygon => it should not.
But why are those tiles not loaded ?

Because the boundaries computed are different from the theorical boundaries : 

<details><summary>code to create figure</summary>

```python
for ax in [ax_pla, ax_ort]:
    c_geom = ax.add_geometries([geom], crs=tiler.crs, color="red", alpha=0.3, zorder=30)
    
geo1 = mpatches.Patch(color='red', label='eroded boundaries')
ax_pla.legend(handles=[geo1, geo2], loc='lower right')
```
</details>

![image](https://user-images.githubusercontent.com/49512274/143694621-6d4f9cb4-20d7-4bad-ae60-8c0a9c74c181.png)

Red boundaries correspond to tiles loaded.
This eroded boundaries are computed [in the `_get_extent_geom` function](https://github.com/SciTools/cartopy/blob/0f60502844fb9bfa0fe339b4b533ac58261a925a/lib/cartopy/mpl/geoaxes.py#L795) with the `.threshold` attribute. 

If we use a lower threshold, we see that there is no more problems

```python
fig = plt.figure(figsize=(10,8), dpi=120)

proj = ccrs.Orthographic(0, 45)
proj.threshold /= 10  # <= lower threshold

ax_pla = fig.add_subplot(2, 1, 1, projection=ccrs.PlateCarree(), frameon=True)
ax_ort = fig.add_axes([0., 0.32, 1, 0.3], projection=proj)

bounds_proj = sgeom.Polygon(proj.boundary)
geom = ax_ort._get_extent_geom(tiler.crs)
img, extent, origin = tiler.image_for_domain(geom, 4)

for ax in [ax_pla, ax_ort]:
    ax.imshow(img, extent=extent, origin=origin, transform=tiler.crs)
    c_bounds = ax.add_geometries([bounds_proj], crs=proj, color="green", alpha=0.3, zorder=40)
    c_geom = ax.add_geometries([geom], crs=tiler.crs, color="red", alpha=0.3, zorder=30)
    
ax_pla.set_title("tiles loaded for ccrs.Orthographic(0, 45) projection, ")

geo2 = mpatches.Patch(color='green', label="ortho boundaries")
geo1 = mpatches.Patch(color='red', label='eroded boundaries')
ax_pla.legend(handles=[geo1, geo2], loc='lower right')
```

![image](https://user-images.githubusercontent.com/49512274/143705499-0d27c52d-0950-406a-8602-0f1d8a1af3ad.png)

[In the code](https://github.com/SciTools/cartopy/blob/0f60502844fb9bfa0fe339b4b533ac58261a925a/lib/cartopy/mpl/geoaxes.py#L793), it is stated that the boundaries are eroded "to avoid transform issues". **What are the cases of those problems ?**

So i ended with this conclusion : 

* The threshold value is too big, **but**
* This threshold problem may be too specific, and don't justify a threshold modification (mainly because it can be solved with `proj.threshold /= 10`  **and**
* threshold modification may introduce transform issues


What do you think about this ? Should we modify the threshold value ?

Thanks

**Comment by greglucas** [comment]
Great investigation! Nice figures to show exactly what is going on.

It looks like the boundary calculation is already pre-eroded in the CRS as well and that could potentially be re-addressed too.
https://github.com/SciTools/cartopy/issues/1154#issuecomment-438976525

I agree with your threshold issue though, right now it seems like it is too large. It is about a factor of 10 larger (relatively) than the PlateCarree threshold of 0.5. So, it seems like perhaps rather than the 2% factor, we could reduce that to something closer to 0.001.
https://github.com/SciTools/cartopy/blob/0f60502844fb9bfa0fe339b4b533ac58261a925a/lib/cartopy/crs.py#L2074

I think the boundary erosion issue would likely take a lot of looking at corner cases and figuring out why tests are failing. It would be nice to remove that, but I think that would  be a pretty big undertaking, so I would focus on the threshold to fix this issue.

**Comment by ludwigVonKoopa** [comment]
Ok i agree too to fix the threshold, i'll submit a PR.

p.s.: how did you put the code line this way ? it quite pretty!

![image](https://user-images.githubusercontent.com/49512274/143771288-2b187dbd-646f-4c60-9a7a-512aa5ae6d29.png)


Thanks

**Comment by greglucas** [comment]
:+1:
If you click on the line number within the code (or hold shift to get multiple lines), there is a "copy permalink" button and then just paste that link in the issue and GitHub will do the nice rendering.

![image](https://user-images.githubusercontent.com/12417828/143773191-220f2bdb-e30e-467e-bbec-6b9323a68c69.png)

**Comment by ludwigVonKoopa** [comment]
Oh! Thanks you!
